Documentation/design-docs/hosting-layer-apis.md

```diff
@@ -0,0 +1,145 @@
+    const char_t *resource_search_paths);
+
+int corehost_resolve_component_dependencies(
+    const char_t *component_main_assembly_path,
+    corehost_resolve_component_dependencies_result_fn result)
+```
+
+Resolve dependencies for the specified component.
+* `component_main_assembly_path` - path to the component
+* `result` - callback which will receive the results of the component dependency resolution
+
+See [Component dependency resolution support in host](host-component-dependencies-resolution.md)
+
+``` C
+typedef void(*corehost_error_writer_fn)(const char_t *message);
+
+corehost_error_writer_fn corehost_set_error_writer(corehost_error_writer_fn error_writer)
+```
+
+Set a callback which will be used to report error messages. By default no callback is registered and the errors are written to standard error.
+* `error_writer` - callback function which will be invoked every time an error is reported. When set to `nullptr`, this function unregisters any previously registered callback and the default behaviour is restored.
+
+The return value is the previouly registered callback (which is now unregistered) or `nullptr` if there was no previously registered callback.
+
+The error writer is registered per-thread. On each thread, only one callback can be registered. Subsequent registrations overwrite the previous ones.
+
+#### [Proposed] .NET Core 3.0+
+
+``` C
+typedef void* context_handle;
+
+struct corehost_context_contract
+{
+    size_t version;
+    context_handle instance;
+    int (*get_property)(
+        context_handle instance,
+        const char_t *key,
+        const char_t **value);
+    int (*set_property)(
+        context_handle instance,
+        const char_t *key,
+        const char_t *value);
+    int (*get_properties)(
+        context_handle instance,
+        size_t *count,
+        const char_t **keys,
+        const char_t **values);
+};
+```
+
+Contract for performing operations on an initialized host context.
+* `version` - version of the struct.
+* `instance` - opaque handle to the initialized host context.
+* `get_property` - function pointer for getting a property on the host context.
+  * `key` - key of the property to get.
+  * `value` - pointer to a buffer with the retrieved property value.
+* `set_property` - function pointer for setting a property on the host context.
+  * `key` - key of the property to set.
+  * `value` - value of the property to set. If `nullptr`, the property is removed.
+* `get_properties` - function pointer for getting all properties on the host context.
+  * `count` - size of `keys` and `values`. If the size is too small, it will be populated with the required size. Otherwise, it will be populated with the size used.
+  * `keys` - buffer to populate with the property keys
+  * `values` - buffer to populate with the property values
+
+``` C
+int corehost_initialize_context(const host_interface_t *init, corehost_context_contract *context_contract)
```

Comment:
We also need some kind of `corehost_close_context` or similar to free the context (would be called from `hostfxr_shutdown`).